src/types/function.rs

```diff
@@ -39,6 +76,38 @@ impl PyCFunction {
         )
     }
 
+    /// Create a new function from a closure.
+    pub fn new_closure<F>(f: F, py: Python) -> PyResult<&PyCFunction>
+    where
+        F: Fn(&types::PyTuple, Option<&types::PyDict>) -> PyResult<PyObject>,
+    {
+        let boxed_fn: Box<dyn Fn(&types::PyTuple, Option<&types::PyDict>) -> PyResult<PyObject>> =
+            Box::new(f);
+        let function_ptr = Box::into_raw(Box::new(boxed_fn));
```

Comment:
Please add a comment here why the function is boxed twice.

Reply:
If we used `Box<F>` and made `run_closure` generic on `F` too we might be able to avoid double-boxing at the cost of additional code size. Potentially worth it for the performance improvement?

Reply:
Nice suggestion, indeed using a generic on `F` for `run_closure` removes the need for double boxing or using `dyn` at all. I've just pushed some changes doing this, I imagine that it's at the cost of some slightly larger generated code but avoiding the double indirection seems worth it.